### PR TITLE
Hide parameters of `Adversary`'s `Perturber` from DDP

### PR DESCRIPTION
**Pull request by mzweilin**
# What does this PR do?

* Hide uninitialized parameters from DDP so that we can run adversarial training with DDP.
* Add test examples of adversarial training.
 
## Type of change

Please check all relevant options.

- [ ] Improvement (non-breaking)
- [x] Bug fix (non-breaking)
- [ ] New feature (non-breaking)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

## Testing

Please describe the tests that you ran to verify your changes. Consider listing any relevant details of your test configuration.

- [x] `pytest`
- [x] `CUDA_VISIBLE_DEVICES=0 python -m mart experiment=CIFAR10_CNN_Adv trainer=gpu trainer.precision=16` reports 70%.
- [x] `CUDA_VISIBLE_DEVICES=0,1 python -m mart experiment=CIFAR10_CNN_Adv trainer=ddp trainer.precision=16 trainer.devices=2 model.optimizer.lr=0.2 trainer.max_steps=2925 datamodule.ims_per_batch=256 datamodule.world_size=2` reports 70%.

## Before submitting

- [x] The title is **self-explanatory** and the description **concisely** explains the PR
- [x] My **PR does only one thing**, instead of bundling different changes together
- [ ] I list all the **breaking changes** introduced by this pull request
- [x] I have commented my code
- [ ] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] I have run pre-commit hooks with `pre-commit run -a` command without errors

## Did you have fun?

Make sure you had fun coding 🙃

### CODE REVIEW

**Review by dxoigmn — APPROVED** [review]
LGTM!

File: .github/PULL_REQUEST_TEMPLATE.md [hunk 1]
```diff
@@ -23,8 +23,9 @@ Please check all relevant options.
 
 Please describe the tests that you ran to verify your changes. Consider listing any relevant details of your test configuration.
 
-- [ ] Test A
-- [ ] Test B
+- [ ] `pytest`
+- [ ] `CUDA_VISIBLE_DEVICES=0 python -m mart experiment=CIFAR10_CNN_Adv trainer=gpu trainer.precision=16` reports 70%.
+- [ ] `CUDA_VISIBLE_DEVICES=0,1 python -m mart experiment=CIFAR10_CNN_Adv trainer=ddp trainer.precision=16 trainer.devices=2 model.optimizer.lr=0.2 trainer.max_steps=2925 datamodule.ims_per_batch=256 datamodule.world_size=2` reports 70%.
```
Comment (dxoigmn) [inline .github/PULL_REQUEST_TEMPLATE.md:28]: Why not just `CUDA_VISIBLE_DEVICES=0,1 python -m mart experiment=CIFAR10_CNN_Adv trainer=ddp trainer.precision=16 trainer.devices=2 datamodule.world_size=2
`?

Comment (mzweilin) [reply .github/PULL_REQUEST_TEMPLATE.md]: We don't get speed-up from two-GPUs if we keep the original learning rate and effective batch size.

Comment (mzweilin) [reply .github/PULL_REQUEST_TEMPLATE.md]: I have added the estimated time per epoch for the two experiments.

**Review by dxoigmn — APPROVED** [review]
